src/Storages/MergeTree/RowOrderOptimizer.cpp

```diff
@@ -0,0 +1,162 @@
+#include <Storages/MergeTree/RowOrderOptimizer.h>
```

Comment:
It would be quite interesting to understand which part of the code is really expensive and then optimize this. I'd expect that it is updating the permutation or cardinality estimation - but which of both exactly?.

Perhaps you could run `perf top` on the side and check the hot paths?
(alternatively, logging can also help to understand the runtime of functions, see my other comment)